webrender/src/platform/unix/font.rs

```diff
@@ -228,100 +262,107 @@ impl FontContext {
+
         let result = unsafe { FT_Render_Glyph(slot, render_mode) };
         if result != SUCCESS {
             error!("Unable to rasterize {:?} with {:?}, {:?}", key, render_mode, result);
             return None;
         }
 
-        let dimensions = match self.get_glyph_dimensions_impl(slot, font.render_mode) {
-            Some(val) => val,
-            None => return None,
-        };
-
         let bitmap = unsafe { &(*slot).bitmap };
         let pixel_mode = unsafe { mem::transmute(bitmap.pixel_mode as u32) };
         info!("Rasterizing {:?} as {:?} with dimensions {:?}", key, render_mode, dimensions);
-        // we may be filling only a part of the buffer, so initialize the whole thing with 0
-        let mut final_buffer = Vec::with_capacity(dimensions.width as usize *
-                                                  dimensions.height as usize *
-                                                  4);
 
-        let offset_x = dimensions.left - unsafe { (*slot).bitmap_left };
-        let src_pixel_width = match pixel_mode {
+        let actual_width = match pixel_mode {
             FT_Pixel_Mode::FT_PIXEL_MODE_LCD => {
                 assert!(bitmap.width % 3 == 0);
                 bitmap.width / 3
             },
-            _ => bitmap.width,
-        };
-        // determine the destination range of texels that `bitmap` provides data for
-        let dst_start = cmp::max(0, -offset_x);
-        let dst_end = cmp::min(dimensions.width as i32, src_pixel_width as i32 - offset_x);
-
-        for y in 0 .. dimensions.height {
-            let src_y = y as i32 - dimensions.top + unsafe { (*slot).bitmap_top };
-            if src_y < 0 || src_y >= bitmap.rows as i32 {
-                for _x in 0 .. dimensions.width {
-                    final_buffer.extend_from_slice(&[0xff, 0xff, 0xff, 0]);
-                }
-                continue
+            FT_Pixel_Mode::FT_PIXEL_MODE_MONO |
+            FT_Pixel_Mode::FT_PIXEL_MODE_GRAY => {
+                bitmap.width
             }
-            let base = unsafe {
-                bitmap.buffer.offset((src_y * bitmap.pitch) as isize)
-            };
-            for _x in 0 .. dst_start {
-                final_buffer.extend_from_slice(&[0xff, 0xff, 0xff, 0]);
+            _ => {
+                panic!("Unexpected pixel mode!");
             }
-            match pixel_mode {
-                FT_Pixel_Mode::FT_PIXEL_MODE_MONO => {
-                    for x in dst_start .. dst_end {
-                        let src_x = x + offset_x;
-                        let mask = 0x80 >> (src_x & 0x7);
-                        let byte = unsafe {
-                            *base.offset((src_x >> 3) as isize)
-                        };
+        } as i32;
+
+        let actual_height = bitmap.rows as i32;
+        let top = unsafe { (*slot).bitmap_top };
+        let left = unsafe { (*slot).bitmap_left };
+        let mut final_buffer = vec![0; (actual_width * actual_height * 4) as usize];
```

Comment:
I see how this new size (`actual_width * actual_height`) is now possible, provided your earlier changes to `RasterizedGlyph`. Basically, the metrics and rasterization are decoupled.